File: 2121-intervals-between-identical-elements/2121-intervals-between-identical-elements.cpp
class Solution {
public:
   vector<long long> getDistances(vector<int>& arr) {
        std::vector<long long> res(arr.size(), 0);
        std::unordered_map<int, std::vector<long long>> num_to_indices;
        
        for(long long i=0; i < arr.size(); i++) {
            num_to_indices[arr[i]].push_back(i);
        }
        
        for(auto const& ele: num_to_indices) {
            auto const& indices = ele.second;
            for(int i = 1; i < indices.size(); i++) {
                res[indices[0]] += std::abs(indices[i]-indices[0]);
            }
            for(int i = 1; i < indices.size(); i++) {
			    // Not simplifying this line intentionally for better understanding.
                res[indices[i]] = res[indices[i-1]] - (indices[i]-indices[i-1])*(indices.size()-i-1) + (indices[i]-indices[i-1])*(i-1);
            }
        }
        
        return res;
    }
};

File: 2121-intervals-between-identical-elements/2121-intervals-between-identical-elements.java
class Solution {
   public long[] getDistances(int[] a){
        long[] r = new long[a.length];
        sum(a, 0, 1, r);
        sum(a, a.length - 1, -1, r);
        return r;
    }

    void sum(int[] a, int start, int increment, long[] r){
        long[] count = new long[100_001], sum = new long[100_001];
        for(int i = start, j = 0; 0 <= i && i < r.length; i += increment, j++){
            r[i] += count[a[i]] * j - sum[a[i]];
            count[a[i]]++;
            sum[a[i]] += j;
        }
    }
}